crates/shared/src/order_quoting.rs

```diff
@@ -530,6 +548,78 @@ impl OrderQuoting for OrderQuoter {
     }
 }
 
+/// Handles whether or not quotes need to be verified.
+enum QuoteVerification {
+    /// Quotes get trusted blindly.
+    Never,
+    /// Enforce quotes only when the user has the required sell token balance.
+    /// If a user has the required sell token balance and a quote can still not
+    /// be verified the trade likely involves a weird token that solvers are not
+    /// able to handle correctly.
+    WhenFunded(Arc<dyn BalanceFetching>),
+}
+
+impl QuoteVerification {
+    /// Makes sure a quote was verified according to the configured rule.
+    async fn check(
+        &self,
+        estimate: &Estimate,
+        parameters: &QuoteParameters,
+        sell_amount: U256,
+    ) -> Result<(), CalculateQuoteError> {
+        let Some(verification) = &parameters.verification else {
+            return Ok(()); // no verification requested
+        };
+        if estimate.verified {
+            return Ok(()); // successfully verified
+        }
+        let balance_fetcher = match &self {
+            QuoteVerification::Never => return Ok(()),
+            QuoteVerification::WhenFunded(fetcher) => fetcher.as_ref(),
+        };
+
+        let balance =
+            match Self::get_balance(balance_fetcher, verification, parameters.sell_token).await {
+                Ok(balance) => balance,
+                Err(err) => {
+                    tracing::warn!(?err, "could not fetch balance for verification");
+                    return Err(CalculateQuoteError::QuoteNotVerified);
+                }
+            };
+
+        if balance >= sell_amount {
```

Comment:
I didn't really get why we expect the balance to be exactly less than the sell amount. Could you elaborate a little bit on that?